lib/Sema/CSGen.cpp

```diff
@@ -747,7 +747,73 @@ namespace {
     
     favorCallOverloads(expr, CS, isFavoredDecl);
   }
-  
+
+  /// If \p expr is a call and that call contains the code completion token,
+  /// return the expressions of all arguments after the code completion token.
+  /// Otherwise, returns an empty vector.
+  /// Asssumes that we are solving for code completion.
+  std::vector<Expr *>
+  getArgumentsAfterCodeCompletionToken(Expr *expr, ConstraintSystem &CS) {
+    assert(CS.isForCodeCompletion());
+
+    /// Don't ignore the rhs argument if the code completion token is the lhs of
+    /// an operator call. Main use case is the implicit `<complete> ~= $match`
+    /// call created for pattern matching, in which we need to type-check
+    /// `$match` to get a contextual type for `<complete>`
+    if (isa<BinaryExpr>(expr)) {
+      return {};
+    }
+
+    auto args = expr->getArgs();
+    auto argInfo = getCompletionArgInfo(expr, CS);
+    if (!args || !argInfo) {
+      return {};
+    }
+
+    std::vector<Expr *> IgnoredArguments;
+
+    for (auto argIndex : indices(*args)) {
+      if (argInfo->isBefore(argIndex)) {
+        IgnoredArguments.push_back(args->get(argIndex).getExpr());
+      }
+    }
+
+    return IgnoredArguments;
+  }
+
+  /// Assuming that we are solving for code completion, assign \p expr a fresh
+  /// and unconstrained type variable as its type.
+  void setTypeForArgumentIgnoredForCompletion(Expr *expr, ConstraintSystem &CS) {
+    assert(CS.isForCodeCompletion());
+
+    /// Creates a fresh type variable that can bind to anything
+    auto createFreshTypeVariable = [&CS](ConstraintLocator *loc) -> Type {
+      return CS.createTypeVariable(loc,
+                                   TVO_CanBindToLValue | TVO_CanBindToInOut |
+                                   TVO_CanBindToNoEscape | TVO_CanBindToHole);
+    };
+
+    if (auto closure = dyn_cast<ClosureExpr>(expr)) {
```

Comment:
There is a method for this already - `inferClosureType` which should do the right thing since parameters are marked as TVO_CanBindToHole by default.

Reply:
Ah, nice. I needed to add another parameter to `inferClosureType` so that the result can also bind to a hole when ignoring arguments, but it’s definitely nicer than my hand-rolled implementation.